### ISSUE TITLE
kv_test occasionally fails with a seg fault

### ISSUE THREAD

**Issue by eddyashton** [microsoft/CCF#755]
https://dev.azure.com/MSRC-CCF/CCF/_build/results?buildId=4363&view=logs&j=f64f962d-f92a-53ce-6755-9d73615a3163&t=f2f25915-3061-5edf-be3d-b5d8edbdb9e4&l=94

So far I've only seen this in the CI. We need to establish whether this is an issue with the KV contention test code, or something deeper in the KV itself. We should see if any thread-safety tooling/linters can spot a potential violation here (currently ASAN has no problem).

**Comment by eddyashton** [comment]
This can block PRs. Azure Pipelines lists 8 failures of this test from 1300 runs in the past 14 days.

**Comment by achamayou** [comment]
We should have a look at setting up https://github.com/google/sanitizers/wiki/ThreadSanitizerAlgorithm at least for this particular test if not for all of them.

**Comment by eddyashton** [comment]
Good call, this has thrown up 2 issues with `kv_test` straight away. I'll attach some verbose error messages below, but in short:
- `loggers()` in `logger.h` isn't threadsafe due to its use of function-statics
- `swap_private_maps()` in `kv.h` may lead to a `kv::Map` being destructed while it still holds a mutex

I'll look at fixes for these, and then at running ThreadSanitizer over more of the code (sadly address san and thread san are mutually exclusive, compile errors if we try to use both, so we may need to choose which tests get which sanitizers).

[logging_statics.txt](https://github.com/microsoft/CCF/files/4135366/logging_statics.txt)
[map_swaps.txt](https://github.com/microsoft/CCF/files/4135367/map_swaps.txt)

**Comment by achamayou** [comment]
```
[doctest] run with "--help" for options
===============================================================================
../src/kv/test/kv_contention.cpp:17:
TEST SUITE: concurrency
TEST CASE:  Concurrent kv access

../src/kv/test/kv_contention.cpp:17: FATAL ERROR: test case CRASHED: SIGSEGV - Segmentation violation signal

2020-02-03T17:12:32.380153Z        100 [fail ] ../src/kv/test/../kv.h:1015          | Could not commit transaction due to conflict
===============================================================================
2020-02-03T17:12:32.380224Z        100 [fail ] ../src/kv/test/../kv.h:1015          | Could not commit transaction due to conflict
[doctest] test cases:      1 |      0 passed2020-02-03T17:12:32.380271Z        100 [fail ] ../src/kv/test/../kv.h:1015          | Could not commit transaction due to conflict
 |      1 failed |     21 skipped
[doctest] assertions: 2020-02-03T17:12:32.380321Z        100 [fail ] ../src/kv/test/../kv.h:1015          | Could not commit transaction due to conflict
     1 |      1 passed |      0 failed |
[doctest] Status: FAILURE!
/mnt/build/2/s/tests/unit_test_wrapper.sh: line 5:  1280 Segmentation fault      (core dumped) LLVM_PROFILE_FILE="${1}.profraw" "./${1}" -nv
```

https://dev.azure.com/MSRC-CCF/CCF/_build/results?buildId=4513&view=ms.vss-test-web.build-test-results-tab&runId=26152&resultId=100003&paneView=debug

**Comment by eddyashton** [comment]
I spent a few days squinting at the reports from ThreadSanitizer, and I've convinced myself that all of the issues it raised were false positives. In particular, the issues it spotted here in the kv_tests aren't issues in practice - the way it stresses `logger.h` can't happen in a real CCF node.

This particular crash has been resolved for unrelated reasons - no repros in 14 days because the offending contentious logging ("Could not commit transaction due to conflict") was reduced to `TRACE` level in #807. I think that's a correct change, so I'm closing this issue.
